### PR TITLE
Issues/91 syncing part 1

### PR DESCRIPTION
**Pull request by aerych**
Refs #91 

This PR is part one of five that will introduce comprehensive syncing support to the app. 

Changes in this PR:
- Support creating a remote draft for a story folder if it needs one. 
- Support for updating a story folder's properties (postID and name) based on the remote draft.
- Support for remove story folders that have been published, marked as spam, or trashed. 
- Introduces a SyncCoordinator. This class is responsible for coordinating batch sync operations in a way that does not conflict with the Reconciler. 
- Adds a bunch of unit tests.

This PR does not: 
- Fully implement sync. The changes are already very large so more work will come in the following PRs. For now assume this is the baseline functionality, but StoryFolder syncing is expected to be incomplete.
- Do any uploading of assets.

Notes: 
- There are fewer real changes in FolderStore than shown by the diff. I moved methods around for better grouping but that code was largely untouched otherwise.   The interesting bits will be in sections for "Syncing related", "Folder Store Retrieval", and the methods for assigning a post ID, and updating sync date.
- I've opted to abandon the flex pattern for the SyncController.  Attempting to leverage it just added additional complexity with meager gain. Instead I'm relying on callback blocks and dispatch queues to ensure work is done when it needs to get done and to report back completion.  I'm not entirely satisfied with the pattern but it's at least a measure better from where Flex was going.  Open to other ideas / suggestions!

### To Test: 
Build with Xcode 11.7 (Xcode 12 support coming in the future.)
Run unit tests.  Ensure they all pass. 

From a clean install 
- Log in.
- Create a new story folder. 
- Go to the photos app and share a photo to your newly created folder.
- Return to the app and observe the console output.  Confirm that you see the sync coordinator started and stopped, and that it ran _after_ the reconciler. 
- Go to your site's and confirm you have a new draft post with the name you gave the story folder. 

Apologies in advance for size of this one!  Could I still trouble you for a peek when you have time @jleandroperez?

### CODE REVIEW

File: Newspack/Newspack/Api/Models/RemotePost.swift [hunk 1]
```diff
@@ -88,3 +88,42 @@ struct RemotePostID {
...
+    /// - Parameter dict: The source dictionary
+    ///
+    init(dict: [String: AnyObject]) {
+        postID = dict[intForKey: "id"]
+        dateGMT = Date.dateFromGMTString(string: dict[stringForKey: "date_gmt"])!
+        link = dict[stringForKey: "link"]
+        modifiedGMT = Date.dateFromGMTString(string: dict[stringForKey: "modified_gmt"])!
+        status = dict[stringForKey: "status"]
+        title = dict[stringAtKeyPath: "title.raw"]
+        titleRendered = dict[stringAtKeyPath: "title.rendered"]
```
Comment (jleandroperez) [inline Newspack/Newspack/Api/Models/RemotePost.swift:114]: This whole subscript trick is wonderful!! ❤️

File: Newspack/Newspack/Api/ServiceRemotes/PostServiceRemote.swift [hunk 2]
```diff
@@ -33,6 +33,41 @@ class PostServiceRemote: ServiceRemoteCoreRest {
...
+        let filter = ["id": ids] as [String: AnyObject]
+        let params = [
+            "_fields": "id,title,status,date_gmt,modified_gmt,_links",
+            "page": page,
+            "per_page": perPage
+        ] as [String: AnyObject]
+        let parameters = params.mergedWith(filter)
+
+        api.GET("posts", parameters: parameters, success: { (response: AnyObject, httpResponse: HTTPURLResponse?) in
+            let array = response as! [[String: AnyObject]]
```
Comment (jleandroperez) [inline Newspack/Newspack/Api/ServiceRemotes/PostServiceRemote.swift]: Q: Would it make sense to switch the `as!` > `as?` ?

File: Newspack/Newspack/Stores/FolderStore.swift [hunk 3]
```diff
@@ -245,39 +217,29 @@ extension FolderStore {
...
-            let folder = context.object(with: objID) as! StoryFolder
-            folder.name = name
-
-            CoreDataManager.shared.saveContext(context: context)
+        // Check for last selected story folder.
+        if let lastFolder = getLastSelectedStoryFolder() {
+            selectStoryFolder(uuid: lastFolder.uuid)
+            return
         }
+        selectStoryFolder(uuid: firstFolder.uuid)
```
Comment (jleandroperez) [inline Newspack/Newspack/Stores/FolderStore.swift]: Nipticky!! You can simplify this snippet:

```
let folder = getLastSelectedStoryFolder() ?? firstFolder
selectStoryFolder(uuid: folder.uuid)
```

**Review by jleandroperez — APPROVED** [review]
✅ Code looks sharp!!
✅ Confirming that I got a new Draft Post with the story name
✅ New Unit Test pass
✅ SyncCoordinator works right after the Reconciler
⚠️ **ShadowTests.testClearShadowAssets** breaks, but it's not new from this PR (breaks in develop!)

I'm assuming the Photo isn't expected to be uploaded, at this stage. Please LMK if that's not the case!!

Looks neat sir!!

:shipit:

File: Newspack/Newspack/Stores/FolderStore.swift (hunk 3 above)
Comment (aerych) [reply Newspack/Newspack/Stores/FolderStore.swift]: Nice! I like it :)

**Comment by aerych** [comment]
> ShadowTests.testClearShadowAssets breaks, but it's not new from this PR (breaks in develop!)

Hmm. This is passing sometimes and sometimes not. Seems like a timing issue with when defaults are synchronized.  Thanks for spotting it!

**Comment by aerych** [comment]
Thank you @jleandroperez !  One last quick peek at the changes and to make sure the test is working properly now?

**Review by jleandroperez — APPROVED** [review]
✅ PR Re-tested, all great!!
✅ Code looks sharp!!

One tiny note: the `testClearShadowAssets` Unit Test is still breaking. Good to ship, since it's not a new thing.

Happy to team up and debug this one anytime sir!! :shipit: on my end!!

**Comment by aerych** [comment]
Thank you @jleandroperez 
That test is a real mystery. I'll sort it out in another PR.

**Pull request merged**